core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -920,6 +927,11 @@ private void acquirePrimaryShardReference(ShardId shardId, String allocationId,
         if (actualAllocationId.equals(allocationId) == false) {
             throw new ShardNotFoundException(shardId, "expected aID [{}] but found [{}]", allocationId, actualAllocationId);
         }
+        final long actualTerm = indexShard.getPrimaryTerm();
```

Comment:
Adding the check to `acquirePrimaryOperationPermit` would be less of a contained change than this (i.e. it would also involve recovery, where would then have to add a term to StartRecoveryRequest, ...). I still plan on exploring that as a follow-up, but thought this to be the easiest way to have a self-contained change that focuses on the TransportReplicationAction / ReplicaRequest.